stdlib/public/core/FloatingPoint.swift

```diff
@@ -1948,9 +1948,30 @@ extension BinaryFloatingPoint {
   /// - Parameter value: A floating-point value to be converted.
   @inlinable
   public init?<Source: BinaryFloatingPoint>(exactly value: Source) {
-    let (value_, exact) = Self._convert(from: value)
-    guard exact else { return nil }
-    self = value_
+    // We define exactness by equality after roundtripping; since NaN is never
+    // equal to itself, it can never be converted exactly.
+    if value.isNaN { return nil }
+    
```

Comment:
I think we can simplify the logic downstream by pulling out zero and infinity here (since both are always exact). WDYT?

Reply:
I left it out because, when specialized, the case of Float-to-Double would be a single 'isNaN' check followed by conversion. Checking for finite nonzero values would have to be written separately anyway, since it wouldn't return nil, which is what happens here already. The only difference is whether it can be short-circuited or not when the concrete types are known; the logic doesn't otherwise get any simpler, and I'd like to be able to short circuit that check when it's not necessary.